nncf/experimental/common/pruning/operations.py

```diff
@@ -0,0 +1,637 @@
+                group = PropagationGroup.join_groups(left[0], right[0])
+                group.close_branch()
+                cls.add_consumer_block(group, node)
+        else:
+            output_mask = node.data.get('output_mask', None)
+
+        node.data['output_mask'] = output_mask
+
+    @classmethod
+    def add_consumer_block(cls, group: PropagationGroup, node: NNCFNode) -> None:
+        """
+        Explicitly adds a consumer block to the given group.
+
+        :param group: the given block group.
+        :param node: node that corresponds to the consumer.
+        """
+        if node.layer_attributes is not None:
+            first_block: PropagationBlock = group.get_blocks()[0]
+            consumer_block = PropagationBlock(
+                MaskProducer(node.node_id),
+                size=first_block.size,
+                offset=first_block.offset,
+                pruning_dimension=1,
+                closed_branches=1
+            )
+            group.add_block(consumer_block)
+
+
+class BatchNormPruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        identity_mask_propagation(node, graph)
+
+
+class GroupNormPruningOp(BasePruningOp):
+    @classmethod
+    def accept_pruned_input(cls, node: NNCFNode) -> bool:
+        # For Instance Normalization
+        return isinstance(node.layer_attributes, GroupNormLayerAttributes) \
+            and node.layer_attributes.num_groups == node.layer_attributes.num_channels
+
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        if cls.accept_pruned_input(node):
+            identity_mask_propagation(node, graph)
+        else:
+            node.data['output_mask'] = None
+
+
+class ConcatPruningOp(BasePruningOp):
+    @classmethod
+    def generate_output_mask(cls, node: NNCFNode, graph: NNCFGraph,
+                             tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> Optional[NNCFTensor]:
+        """
+        Generate output mask from input masks with all None replaced by identity masks.
+        If all input masks is None return None.
+
+        :param node: Node to determine it's sources.
+        :param graph: NNCF graph to work with.
+        :param tensor_processor: Interface with tensor processing methods.
+        :return: Filled input masks.
+        """
+        input_edges = graph.get_input_edges(node)
+        previous_nodes = [edge.from_node for edge in input_edges]
+        input_masks = [input_node.data['output_mask'] for input_node in previous_nodes] #  type: List[NNCFTensor]
+        not_empty_masks = [mask for mask in input_masks if mask is not None]
+        if not not_empty_masks:
+            return None
```

Comment:
This method always returns `None`

Reply:
Yes, will be fixed it the next PR #1680 Currently, it's enough for 3 models.

Reply:
Could you please remove unused code then?